File: java/509. Fibonacci Number/Solution.java
class Solution {
     private Map<Integer, Integer> memo = new HashMap<>(Map.of(0, 0, 1, 1));

     public int fib(int n) {
         if (!memo.containsKey(n)) {
             memo.put(n, fib(n - 1) + fib(n - 2));
         }
         return memo.get(n);
     }

    public int fib(int n) {
        if (n == 0 || n == 1) return n;
        int[] memo = new int[n + 1];
        memo[1] = 1;

        for (int i = 2; i < n + 1; i++) {
            memo[i] = memo[i - 1] + memo[i - 2];
        }

        return memo[n];
    }


}

File: python/509. Fibonacci Number/solution.py
class Solution:
    def fib(self, n: int) -> int:
        memo = {}
        if n == 0 or n == 1:
            return n
        if n not in memo:
            memo[n] = self.fib(n - 1) + self.fib(n - 2)
        return memo[n]

    def fib(self, n: int) -> int:
        if n == 0 or n == 1:
            return n
        memo = [None] * (n + 1)
        memo[0] = 0
        memo[1] = 1

        for i in range(2, n + 1):
            memo[i] = memo[i - 1] + memo[i - 2]
        return memo[n]